src/dd/Approximation.cpp

```diff
@@ -0,0 +1,246 @@
+using Lookup = std::unordered_map<const vNode*, vEdge>;
+/// @brief A node, a flag indicating left or right, and its contribution.
+using Terminal = std::tuple<vNode*, uint16_t, double>;
+/// @brief List of potentially deletable terminal edges.
+using TerminalList = std::forward_list<Terminal>;
+
+/**
+ * @brief Remove terminals with a contribution above the new budget.
+ */
+TerminalList removeAbove(const TerminalList& old, const double budget) {
+  TerminalList upd{};
+  for (const auto& candidate : old) {
+    if (budget >= std::get<2>(candidate)) {
+      upd.emplace_front(candidate);
+    }
+  }
+  return upd;
+}
+
+/**
+ * @brief Search and mark nodes for deletion until the budget 1 - @p fidelity is
+ * exhausted.
+ * @details Uses a prioritized iterative-deepening search.
+ * Iterating layer by layer ensures that each node is only visited once.
+ */
+std::pair<double, Qubit> mark(VectorDD& state, const double fidelity,
+                              struct ApproxMeta* meta) {
+  Layer curr{};
+  curr.emplace(state.p);
+
+  double budget = 1 - fidelity;
+  Qubit min{std::numeric_limits<Qubit>::max()};
+
+  TerminalList candidates{};
+
+  ApproxMeta rmeta{};
+
+  while (budget > 0) {
+    Contributions c; // Stores contributions of the next layer.
+    while (!curr.empty()) {
+      const LayerNode n = curr.top();
+      curr.pop();
+      rmeta.nodesVisited++;
+
+      // If possible, flag a node for deletion and decrease the budget.
+      // If necessary, reset the lowest qubit number effected.
+      if (budget >= n.contribution) {
+        n.ptr->flags = FLAG_DELETE;
+        budget -= n.contribution;
+        min = std::min(min, n.ptr->v);
+        removeAbove(candidates, budget);
+        continue;
+      }
+
+      // Compute the contributions of the next layer.
+      for (std::size_t i = 0; i < RADIX; ++i) {
+        const vEdge& eRef = n.ptr->e[i];
+        const double contribution =
+            n.contribution * ComplexNumbers::mag2(eRef.w);
+
+        if (eRef.isTerminal()) {        // Don't add terminals to the queue.
+          if (budget >= contribution) { // They (potentially) can be deleted.
+            const uint16_t flag = (i == 0 ? FLAG_LEFT : FLAG_RIGHT);
+            candidates.emplace_front(n.ptr, flag, contribution);
+          }
+          continue;
+        }
+
+        c[eRef.p] += contribution;
+      }
+    }
+
+    Layer next{}; // Prioritize nodes for next iteration.
+    for (auto& [n, contribution] : c) {
+      next.emplace(n, contribution, budget);
+    }
+
+    if (next.empty()) { // Break early. Avoid std::move.
+      break;
+    }
+
+    curr = std::move(next);
+  }
+
+  // Lastly, check if any terminals can be deleted.
+  while (!candidates.empty()) {
+    const auto [n, flag, contribution] = candidates.front();
+    candidates.pop_front();
+    if (budget >= contribution) {
+      n->flags = FLAG_DELETE + flag;
+      budget -= contribution;
+      min = std::min(min, n->v);
+    }
+  }
```

Comment:
There is no real need to manipulate the list here. It can simply be traversed in a read-only fashion from beginning to end (or until the budget is exhausted).